gh-115627: Fix PySSL_SetError handling SSL_ERROR_SYSCALL

keepworking opened the pull request:
<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
gh-NNNNN: Summary of the changes made
```

Where: gh-NNNNN refers to the GitHub issue number.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `main`.

-->

The SSL_write() and SSL_read() used in python 3.9 were modified from python 3.10 to SSL_write_ex() and SSL_read_ex(), and the return value was fixed to 0 on error.

I modified it so that OSError can be returned correctly, not EOF, when retval is 0.

<!-- gh-issue-number: gh-115627 -->
* Issue: gh-115627
<!-- /gh-issue-number -->

keepworking committed:
gh-115627: Fix PySSL_SetError handling SSL_ERROR_SYSCALL

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

serhiy-storchaka commented on Modules/_ssl.c:
```diff
@@ -645,11 +645,11 @@ PySSL_SetError(PySSLSocket *sslsock, int ret, const char *filename, int lineno)
                     type = state->PySSLEOFErrorObject;
                     errstr = "EOF occurred in violation of protocol";
-                } else if (s && ret == -1) {
+                } else {
```
Should not it be?
```suggestion
                } else if (s) {
```

keepworking replied:
@serhiy-storchaka 
Yes, you are right. I was mistaken that Py_None would be the same as NULL. I will revise as your advice.

serhiy-storchaka commented on Modules/_ssl.c:
```diff
@@ -645,11 +645,11 @@ PySSL_SetError(PySSLSocket *sslsock, int ret, const char *filename, int lineno)
             if (e == 0) {
                 PySocketSockObject *s = GET_SOCKET(sslsock);
-                if (ret == 0 || (((PyObject *)s) == Py_None)) {
+                if (((PyObject *)s) == Py_None) {
```
I think that the `ret` parameter should be removed. `PySSL_SetError()` is a static function, so we can do this.

keepworking replied:
@serhiy-storchaka 
Thank you, I removed the 'intret' which is not used as you advised.

serhiy-storchaka commented on Misc/NEWS.d/next/Library/2024-02-18-09-50-31.gh-issue-115627.HGchj0.rst:
```diff
@@ -0,0 +1,2 @@
+Fix :c:func:`PySSL_SetError` : Modify retval handling logic for handling
```
`PySSL_SetError` is not user visible. It is a static function. Please describe the change in terms of the user visible behavior.

keepworking replied:
Thank you, I've rewritten it to explain how it changed with an error that the user can check.

keepworking committed:
Fix PySSL_SetError function argument (remove int ret)

keepworking committed:
Fix PySSL_SetError function for handling socket variable (able to handle s == NULL)

keepworking committed:
Fix NEWS.d (make it user visible)

serhiy-storchaka commented on Misc/NEWS.d/next/Library/2024-02-18-09-50-31.gh-issue-115627.HGchj0.rst:
```diff
@@ -1,2 +1,2 @@
-Fix :c:func:`PySSL_SetError` : Modify retval handling logic for handling
-SSL_ERROR_SYSCALL.
+FIX the ssl module error handling of connection terminate by peer.
+it throws OSError instead of EOFError.
```
```suggestion
Fix the :mod:`ssl` module error handling of connection terminate by peer.
It now throws an OSError with the appropriate error code instead of an EOFError.
```

keepworking replied:
Thank you for the correction!

serhiy-storchaka commented on Lib/test/test_ssl.py:
```diff
@@ -2397,16 +2397,19 @@ def run(self):
+                        # OpenSSL 1.1.1 sometimes raises
+                        # ConnectionResetError when connection is not
+                        # shut down gracefully.
+                        print(
```
Only print if `support.verbose` is true. Tests should not produce additional output in the quiet mode.

keepworking replied:
Thank you, as you advised, the print has been modified to work according to the existing conditions.

keepworking committed:
Fix test_ssl.py print server log only verbose mode

keepworking committed:
Fix PySSL_SetError function do not reference socketobject for handling error

keepworking committed:
Update Misc/NEWS.d/next/Library/2024-02-18-09-50-31.gh-issue-115627.HGchj0.rst

Co-authored-by: Serhiy Storchaka <storchaka@gmail.com>

serhiy-storchaka reviewed (commented):
Thank you for your contribution @keepworking.

I checked that the code correctly does what you intended, but I am not a SSL expert and does not know if it is the right thing. So we will wait for review from SSL experts.

encukou commented on Lib/test/test_ssl.py:
```diff
@@ -3167,7 +3169,7 @@ def test_wrong_cert_tls13(self):
             with self.assertRaisesRegex(
                 ssl.SSLError,
-                'alert unknown ca|EOF occurred|TLSV1_ALERT_UNKNOWN_CA'
+                'alert unknown ca|EOF occurred|TLSV1_ALERT_UNKNOWN_CA|closed by the remote host'
```
This can now raise `ConnectionResetError` too (` [WinError 10054] An existing connection was forcibly closed by the remote host` or `[Errno 104] Connection reset by peer`)
```suggestion
                'alert unknown ca|EOF occurred|TLSV1_ALERT_UNKNOWN_CA|closed by the remote host|Connection reset by peer'
```

keepworking replied:
Thank you for your kind correction! I tested the changes you suggested at home and confirmed that they work correctly. 😄

encukou commented on Lib/test/test_ssl.py:
```diff
@@ -4532,8 +4534,8 @@ def msg_cb(conn, direction, version, content_type, msg_type, data):
                 with self.assertRaisesRegex(
-                    ssl.SSLError,
-                    '(certificate required|EOF occurred)'
+                    (ssl.SSLError, OSError),
```
`ssl.SSLError` is a subclass of `OSError`.
```suggestion
                    OSError,
```

encukou commented on Lib/test/test_ssl.py:
```diff
@@ -3167,7 +3169,7 @@ def test_wrong_cert_tls13(self):
             s.connect((HOST, server.port))
             with self.assertRaisesRegex(
                 ssl.SSLError,
```
```suggestion
                OSError,
```

encukou reviewed (commented):
I'm now also convinced that the fix is correct. 

<details>

<summary>(Not being familiar with the code, I went the hard way)</summary>

I read the man pages of all functions for which `PySSL_SetError` is called, for OpenSSL 3.0 and 1.1.1. In all cases, `SSL_get_error` (and no other mechanism) should be called to get the error; our `PySSL_SetError` indeed shouldn't need the return value. It seems that the meaning of the return value is function-specific; the general `PySSL_SetError` shouldn't do things like `(asert ret <= 0)`.

In all cases, `SSL_get_error` *is* called via `_PySSL_errno(...)` and set on `self->err.ssl`, before `PySSL_SetError` is called.

- SSL_do_handshake, if (ret < 1)
- SSL_write_ex, if (retval == 0)
- SSL_pending, if (count < 0)
    (this should not happen: man page does not mention the function returning <0, nor any way it can signal an error)
- SSL_read_ex, if (retval == 0)
- SSL_shutdown, if (ret < 0)

</details>

However, the function now raises OSError rather than SSLError more often, and the tests need to reflect that.

Could you address the CI failures? After that let's also test on buildbots.

keepworking committed:
Apply suggestions from code review

- merge ssl.SSLError and OSError as OSError
- append assertRaisesRegex for test_wrong_cert_tls13 ('Connection reset by peer')

Co-authored-by: Petr Viktorin <encukou@gmail.com>

keepworking committed:
Modify assertRaisesRegex for ThreadedEchoServer (append Connection reset by peer)
